StorageUtility/Windows/Device.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+<License>
+Copyright 2015 Virtium Technology
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http ://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+</License>
+*/
+#include "Device.h"
+#include "StorageUtility.h"
+
+namespace vtStor
+{
+
+cDevice::cDevice(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData)
+{
+    AllocateMemories(DevInfoData, DevInterfaceData, DevDetailData, SizeOfDevDetailData);
+    CopyMemories(DevInfoData, DevInterfaceData, DevDetailData);
+}
+
+cDevice::~cDevice()
+{
+    DeallocateMemories();
+}
+
+void cDevice::Data(std::unordered_map<eDeviceDataType, void*>& Data)
+{
+    Data.insert({ eDeviceDataType::PSP_DEVINFO_DATA, m_DevInfoData });
+    Data.insert({ eDeviceDataType::PSP_DEVICE_INTERFACE_DATA, m_DevInterfaceData });
+    Data.insert({ eDeviceDataType::PSP_INTERFACE_DEVICE_DETAIL_DATA, m_DevDetailData });
+}
+
+DeviceHandle cDevice::Handle()
+{
+    eErrorCode errorCode;
+    DeviceHandle deviceHandle;
+
+    errorCode = GetStorageDeviceHandle(m_DevDetailData->DevicePath, deviceHandle);
+    if (eErrorCode::None != errorCode)
+    {
+        throw std::exception("Failed to create device handle", (int)errorCode);
+    }
+    return deviceHandle;
+}
+
+eErrorCode cDevice::GetStorageDeviceHandle(const String& DevicePath, HANDLE& Handle)
+{
+    Handle =
+        CreateFile(
+        DevicePath.c_str(),
+        GENERIC_READ | GENERIC_WRITE,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        NULL,
+        OPEN_EXISTING,
+        FILE_ATTRIBUTE_NORMAL,
+        NULL
+        );
+
+    if (INVALID_HANDLE_VALUE == Handle)
+    {
+        return(eErrorCode::Io);
+    }
+
+    return(eErrorCode::None);
+}
+
+void cDevice::AllocateMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData)
+{
+    m_DevInfoDataHeap = GetProcessHeap();
+    m_DevInterfaceDataHeap = GetProcessHeap();
+    m_DevDetailDataHeap = GetProcessHeap();
```

Comment:
GetProcessHeap() can be call inline with the memory functions. No need to save three duplicate handles.